### DISCUSSION TITLE
ValueObjects com o .NET 7+ e Entity Framework Core 7+

### DISCUSSION THREAD

**Discussion by alissonpratesperes** [balta-io/forum#376] [Tira Dúvidas] [answered]
Estou finalizando o desenvolvimento de uma App, como um trabalhinho de conclusão da Carreira Desenvolvedor Backend .NET, com uma ideia minha. E trago um questionamento que possa vir a ajudar qualquer outro que também o tiver. Aqui estou usando, dentre vários Design Patterns, o DDD, e como procurei implementar um pouco de cada feature coberta no curso, travei 3 dias num problema quanto à implementação de ValueObjects...
Seguindo a premissa dos ValueObjects, uma das minhas Entidades, possui as duas coordenadas de localização que serão usadas posteriormente no meu Front-End em um componente de mapas... Então resolvi abstrair para um ValueObject e encapsular elas, então criei o ValueObject com as duas propriedades, alterei no mapper dessa entidade (usando o OwnsOne com loop, explicitando que as duas propriedades estão encapsuladas nesse VO), e mesmo assim minha App não estava subindo, com o erro: Cannot bind 'location' in 'Entity(... Coordinate location, ...)' Note that only mapped properties can be bound to constructor parameters. Navigations to related entities, including references to owned types, cannot be bound. - onde Coordinate é meu VO e 'location' a propriedade tipada com ele. Eis que me vi com um problema: o .NET, mais especificamente o Entity Framework Core (estou usando Code First com o SQLite), trava o VO também como entidade e as propriedades que retornam uma "Entidade" tipo são tratadas como propriedades de navegação, só que o meu VO não é uma entidade (a minha Entity sim, herda de Entity - fiz igual ao curso).
Então basicamente tive que alterar minha model principal, deixar o Construtor Publico dela (que contém todos os campos) com o VO, mas acima criar um Construtor Privado da mesma entidade mas sem o campo Coordinate location.

O meu fluxo ficou esse:

1 - Classe abstrata ValueObject - apenas como identificador assim como no curso;
2 - Classe VO Coordinate com lat + long;
3 - Model com o tipo Coordinate na propriedade 'location';
4 - Public Constructor com todos os campos incluindo o "Coordinate location";
5 - Private Constructor com todos os campos menos o "Coordinate location";
6 - No Mapper, usado o OwnsOne com loop para as duas propriedades dentro do VO;
7 - Handlers: new Coordinate(lat, long) no campo da request relacionado à location.
8 - Migration - sim eu achei que não precisava fazer que não alterava no DB, mas altera as colunas da tabela.

O que eu quero saber é se essa abordagem é válida ou é uma gambiarra. Eu sinceramente achei que não seria tão complicado assim de implementar, que seria só criar o VO, adicionar na Model e já era, mas não, todo esse fluxo foi necessário. Fiz correto? Apesar de muita pesquisa, foi à essa conclusão que cheguei, está funcionando mas sinto que isso é uma gambiarra. Diziam que o EF Core 7+ talvez suportaria os VOs nos Construtores, mas pelo visto não.

Obri professor, C# é nota 10!

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @alissonpratesperes, como vai?
No caso, não seria apenas colocar o contrutor como `protected` para atender o EF como mostro nos cursos?

De resto me parece OK!

**Reply by alissonpratesperes** [reply]
Tudo certo e contigo?!

Nesse caso igual eu teria os dois construtores, mas ao invés do Private, usaria o modificador Protected, isso? Mas igual teria o Protected sem o campo mencionado e o Public com o Campo.. Isso?
